src/configs/cpp.rs

```diff
@@ -0,0 +1,43 @@
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Deserialize, Serialize)]
+#[cfg_attr(
+    feature = "config-schema",
+    derive(schemars::JsonSchema),
+    schemars(deny_unknown_fields)
+)]
+#[serde(default)]
+pub struct CppConfig<'a> {
+    pub format: &'a str,
+    pub version_format: &'a str,
+    pub style: &'a str,
+    pub symbol: &'a str,
+    pub disabled: bool,
+    pub detect_extensions: Vec<&'a str>,
+    pub detect_files: Vec<&'a str>,
+    pub detect_folders: Vec<&'a str>,
+    pub commands: Vec<Vec<&'a str>>,
+}
```

Comment:
What do you think about handling it like that?
```suggestion
#[serde(transparent)]
pub struct CppConfig<'a>(CConfig<'a>)
```

Reply:
I tried implementing it the way you suggested but I am facing lifetime related issues in deserialize,

impl:

```rust
use serde::{Deserialize, Serialize};
use crate::configs::c::CConfig;

#[derive(Clone, Deserialize, Serialize)]
#[cfg_attr(
    feature = "config-schema",
    derive(schemars::JsonSchema),
    schemars(deny_unknown_fields)
)]
#[serde(transparent)]
pub struct CppConfig<'a>(CConfig<'a>);

impl Default for CppConfig<'_> {
    fn default() -> Self {
        CppConfig(CConfig {
            format: "via [$symbol($version(-$name) )]($style)",
            version_format: "v${raw}",
            style: "149 bold",
            symbol: "C++ ",
            disabled: false,
            detect_extensions: vec![
                "cc", "cp", "cxx", "cpp", "CPP", "c++", "C", "hh", "H", "hp", "hxx", "hpp", "HPP",
                "h++", "tcc",
            ],
            detect_files: vec![],
            detect_folders: vec![],
            commands: vec![
                vec!["c++", "--version"],
                vec!["g++", "--version"],
                vec!["clang++", "--version"],
            ],
        })
    }
}
```

error:

```console
error: lifetime may not live long enough
  --> src/configs/cpp.rs:4:17
   |
4  | #[derive(Clone, Deserialize, Serialize)]
   |                 ^^^^^^^^^^^
   |                 |
   |                 lifetime `'de` defined here
   |                 associated function was supposed to return data with lifetime `'a` but it is returning data with lifetime `'de`
...
11 | pub struct CppConfig<'a>(CConfig<'a>);
   |                      -- lifetime `'a` defined here
   |
   = help: consider adding the following bound: `'de: 'a`
   = note: this error originates in the derive macro `Deserialize` (in Nightly builds, run with -Z macro-backtrace for more info)
```

I tried doing as compiler suggested. 
```rust
#[serde(transparent)]
pub struct CppConfig<'a, 'de : 'a>(CConfig<'a>);
```
but then I get the error 
```console
cannot deserialize when there is a lifetime parameter called 'de [macro-error]
rustc: cannot deserialize when there is a lifetime parameter called 'de
```

using,
```console
cargo expand configs::cpp --lib
```
I can see the error likely happens due to,
```rust
    #[doc(hidden)]
    #[allow(non_upper_case_globals, unused_attributes, unused_qualifications)]
    const _: () = {
        #[allow(unused_extern_crates, clippy::useless_attribute)]
        extern crate serde as _serde;
        #[automatically_derived]
        impl<'de, 'a> _serde::Deserialize<'de> for CppConfig<'a> {
            fn deserialize<__D>(
                __deserializer: __D,
            ) -> _serde::__private::Result<Self, __D::Error>
            where
                __D: _serde::Deserializer<'de>,
            {
                _serde::__private::Result::map(
                    _serde::Deserialize::deserialize(__deserializer),
                    |__transparent| CppConfig { 0: __transparent },
                )
            }
        }
    };
```

any suggestions? I am kind of stuck here...

Reply:
I think I might have resolved it?

```rust
use serde::{Deserialize, Serialize, Deserializer};
use crate::configs::c::CConfig;

#[derive(Clone, Serialize)]
#[cfg_attr(
    feature = "config-schema",
    derive(schemars::JsonSchema),
    schemars(deny_unknown_fields)
)]
#[serde(transparent)]
pub struct CppConfig<'a>(CConfig<'a>);

impl<'de> Deserialize<'de> for CppConfig<'de>
{
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
            Result::map(
                Deserialize::deserialize(deserializer),
                |transparent| CppConfig { 0: transparent },
            )
    }
}

impl Default for CppConfig<'_> {
    fn default() -> Self {
        CppConfig(CConfig {
            format: "via [$symbol($version(-$name) )]($style)",
            version_format: "v${raw}",
            style: "149 bold",
            symbol: "C++ ",
            disabled: false,
            detect_extensions: vec![
                "cc", "cp", "cxx", "cpp", "CPP", "c++", "C", "hh", "H", "hp", "hxx", "hpp", "HPP",
                "h++", "tcc",
            ],
            detect_files: vec![],
            detect_folders: vec![],
            commands: vec![
                vec!["c++", "--version"],
                vec!["g++", "--version"],
                vec!["clang++", "--version"],
            ],
        })
    }
}
```

Reply:
never mind it came back for 
```rust
    #[serde(borrow)]
    cpp: cpp::CppConfig<'a>,
```
in FullConfig.

Reply:
finally figured out...
```
#[derive(Clone, Deserialize, Serialize)]
#[cfg_attr(
    feature = "config-schema",
    derive(schemars::JsonSchema),
    schemars(deny_unknown_fields)
)]
#[serde(transparent)]
pub struct CppConfig<'a>(#[serde(borrow)] CConfig<'a>);
```

just had to do serde borrow.
sorry for spamming... very new to rust.

Reply:
@davidkna 

please take a look again. I have made the suggested changes.

Reply:
@davidkna 
I have also updated the docs in the meantime... Pls take a look at them as well.